src/ziplist.c

```diff
@@ -557,7 +557,7 @@ void zipSaveInteger(unsigned char *p, int64_t value, unsigned char encoding) {
         memcpy(p,&i16,sizeof(i16));
         memrev16ifbe(p);
     } else if (encoding == ZIP_INT_24B) {
-        i32 = value<<8;
```

Comment:
If value is negative, it's UB. (Negative integer shift)

Reply:
really? now any shift on a signed integer is UB?
this is BS, left shift doesn't care of the signness of the variable, or the content of the bits.
it could overflow and become positive or negative according to the contents of the bits, but what's undefined about that?

anyway, considering i know how this code is compiled to assembly, i don't see how this could have caused bugs.
i.e. it would move the bits to the upper part of the integer, possibly flip the bytes in BE, and then copy them.
let me know if you think i'm wrong.

Reply:
If -fwrapv used, I think no need to say whether Redis works with 2's compliment or not, compiler will generate correct code for any CPU if I'm not mistaken. AFAIK Linux uses -fwrapv and works everywhere :)

Reply:
unlike the unaligned pointer access, the majority of the changes in this PR for integer overflows seem to me that are ok to be merged (they don't introduce any ugly code or efficiency issues), and the main point of the lengthy discussion for was to rant about it being BS (because it's fun to discuss).

I do like the idea of getting rid of the ugly code in #8910 and the few other places that do that.
however, i don't like to depend on specific compiler flags that may be missing in some odd/old compilers (unless we use these flags only to silence the sanitizer in it's CI build).
So the question is if we drop these changes and add a compilation flag just for one compiler, if we're not taking a risk that the other compilers will decide to mess up the perfectly good code, just because in theory it relies on UB.

am i right? if i am, then we can't do that.

Reply:
It's also possible to disable certain kinds of UB warnings by adding `-fno-sanitize=shift` after `-fsanitize=undefined` to disable this warning.